runtime/src/bank.rs

```diff
@@ -8131,26 +8170,68 @@ impl Bank {
             if let Some(new_account) = self.get_account_with_fixed_root(new_address) {
                 datapoint_info!(datapoint_name, ("slot", self.slot, i64));
 
-                // Burn lamports in the old account
-                self.capitalization
-                    .fetch_sub(old_account.lamports(), Relaxed);
-
-                // Transfer new account to old account
-                self.store_account(old_address, &new_account);
-
-                // Clear new account
-                self.store_account(new_address, &AccountSharedData::default());
+                self.replace_account(old_address, new_address, Some(&old_account), &new_account);
 
                 // Unload a program from the bank's cache
                 self.loaded_programs_cache
                     .write()
                     .unwrap()
                     .remove_programs([*old_address].into_iter());
+            }
+        }
+    }
+
+    /// Use to replace an empty account with a program by feature activation
+    #[allow(dead_code)]
+    fn replace_empty_account_with_upgradeable_program(
+        &mut self,
+        old_address: &Pubkey,
+        new_address: &Pubkey,
+        datapoint_name: &'static str,
+    ) {
+        // Must be attempting to replace an empty account with a program
+        // account _and_ data account
+        if let Some(new_account) = self.get_account_with_fixed_root(new_address) {
+            let (old_data_address, _) = Pubkey::find_program_address(
+                &[old_address.as_ref()],
+                &bpf_loader_upgradeable::id(),
+            );
+            let (new_data_address, _) = Pubkey::find_program_address(
+                &[new_address.as_ref()],
+                &bpf_loader_upgradeable::id(),
+            );
+            if let Some(new_data_account) = self.get_account_with_fixed_root(&new_data_address) {
+                datapoint_info!(datapoint_name, ("slot", self.slot, i64));
 
-                self.calculate_and_update_accounts_data_size_delta_off_chain(
-                    old_account.data().len(),
-                    new_account.data().len(),
+                let data = old_data_address.as_ref().to_vec();
+                let lamports = self.get_minimum_balance_for_rent_exemption(data.len());
```

Comment:
What exactly do you mean by "create new"?

`get_minimum_balance_for_rent_exemption` just calculates the rent due, and the change in capitalization is subtracting them from the new account's lamports, which is destroyed in the process, effectively burning the remaining lamports of the new account after rent is subtracted.

In other words, I set this up so no lamports would be created, but instead taken from the new account which is about to be wiped

Reply:
Sorry, I meant to say "has the possibility to create new lamports if the new program account doesn't have any", but that should never happen because of what you do later

Reply:
Ohh yep I added a commit for tighter lamport checks, pushing soon